### PR TITLE
Use enum for method rather than string compares

### PR DESCRIPTION
**Pull request by benaadams**
For GET method `IsHead` test is performed on the string `Method` which since `GET` isn't reference equal to `HEAD` drops through to `String.Equals` -> `OrdinalIgnoreCaseComparer.Equals`

However the Parser starts with it as an Enum, which then gets converted to a string; so it can just be tested with the enum.

Rebased https://github.com/aspnet/KestrelHttpServer/pull/2027

Resolves: https://github.com/aspnet/KestrelHttpServer/issues/2020

### CODE REVIEW

File: src/Kestrel.Core/Internal/Http/Http1Connection.cs [hunk 1]
```diff
@@ -182,13 +182,16 @@ public void OnStartLine(HttpMethod method, HttpVersion version, Span<byte> targe
...
+            if (method == HttpMethod.Custom)
+            {
+                _customMethod = customMethod.GetAsciiStringNonNullCharacters();
+            }
+
             _httpVersion = version;
 
             Debug.Assert(RawTarget != null, "RawTarget was not set");
-            Debug.Assert(Method != null, "Method was not set");
+            Debug.Assert(((IHttpRequestFeature)this).Method != null, "Method was not set");
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http/Http1Connection.cs]: Why this change?

Comment (davidfowl) [reply src/Kestrel.Core/Internal/Http/Http1Connection.cs]: Ah, because method is an enum.

Comment (benaadams) [reply src/Kestrel.Core/Internal/Http/Http1Connection.cs]: To get the string version

**Comment by benaadams** [comment]
OSX
```
Failed   RegisterAddresses_IPv6_Success(addressInput: "http://*:0/", testUrls: ["http://127.0.0.1", "http://[::1]"])
Error Message:
 System.Net.Sockets.SocketException : Operation timed out
Stack Trace:
   at Microsoft.AspNetCore.Testing.HttpClientSlim.GetSocket(Uri requestUri)
   at Microsoft.AspNetCore.Testing.HttpClientSlim.GetStream(Uri requestUri, Boolean validateCertificate)
   at Microsoft.AspNetCore.Testing.HttpClientSlim.GetStringAsync(Uri requestUri, Boolean validateCertificate)
   at Microsoft.AspNetCore.Testing.HttpClientSlim.GetStringAsync(String requestUri, Boolean validateCertificate)
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.AddressRegistrationTests.RegisterAddresses_Success(String addressInput, String[] testUrls, Int32 testPort) in /_/test/Kestrel.FunctionalTests/AddressRegistrationTests.cs:line 179
   at Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.AddressRegistrationTests.RegisterAddresses_IPv6_Success(String addressInput, String[] testUrls) in /_/test/Kestrel.FunctionalTests/AddressRegistrationTests.cs:line 121
```

**Comment by benaadams** [comment]
Reworded commit to retrigger CI

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs [hunk 2]
```diff
@@ -177,6 +177,110 @@ internal static unsafe HttpMethod GetKnownMethod(byte* data, int length, out int
...
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            var length = value.Length;
+            if (length == 0)
+            {
+                throw new ArgumentException(nameof(value));
+            }
+
+            var firstChar = value[0];
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Why this implementation if its no performance sensitive? Just use the http abstractions methods

Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Is used by http2 so may be perf sensitive :-/

Also are aspnet dlls ngen'd in the newest versions? Does mean none of the methods can be inlined if so.

Did you mean
```csharp
if (HttpMethods.IsGet(value))
{
    method = HttpMethod.Get;
    return true;
}
```

File: src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs [hunk 3]
```diff
@@ -90,8 +90,21 @@ string IHttpRequestFeature.Scheme
...
-            get => Method;
-            set => Method = value;
+            get
+            {
+                if (_customMethod != null)
+                {
+                    return _customMethod;
+                }
+
+                _customMethod = HttpUtilities.MethodToString(Method);
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: The old code did `HttpUtilities.MethodToString(method) ?? string.Empty`. I'd either keep that or have MethodToString throw instead of return null neither of which should ever actually happen in practice.

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs (hunk 2 above)
Comment (halter73) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: How is this used by HTTP/2? If this is only hit by applications setting the request method themselves, I too would prefer a more readable slightly slower version of this method.

Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: HTTP/2 goes via string header
```csharp
((IHttpRequestFeature)this).Method = RequestHeaders[":method"];
```
Though it probably shouldn't use an interface cast for it https://github.com/dotnet/coreclr/issues/16198

Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: e.g. first line of `Http2Stream.TryParseRequest(ReadResult result, out bool endConnection)`

File: src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs (hunk 3 above)
Comment (Tratcher) [inline src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: Why store this in _costomMethod? It's a static lookup. You store null for Set.

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs (hunk 2 above)
Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Improved the `TryParseRequest` not to cast via interface and use method directly

File: src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs (hunk 3 above)
Comment (benaadams) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: done

File: src/Kestrel.Core/Internal/Http/HttpProtocol.cs [hunk 4]
```diff
@@ -324,7 +325,8 @@ public void Reset()
             MaxRequestBodySize = ServerOptions.Limits.MaxRequestBodySize;
             AllowSynchronousIO = ServerOptions.AllowSynchronousIO;
             TraceIdentifier = null;
-            Method = null;
+            Method = HttpMethod.Get;
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.cs]: Added

File: src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs (hunk 3 above)
Comment (benaadams) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: `GetKnownMethod` returns an `HttpMethod` enum for known methods, but not the string so null.

The field is there anyway for a non-known method, so once you've looked it up once, might as well store it if its checked again rather than doing another lookup,

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: So why didn't you store it in the Set code path?

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: If you're going to use it for both then call it `_methodText`

Comment (benaadams) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: Ahh.. that's a very good point 😄

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs (hunk 2 above)
Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Also changed to use the `HttpMethods` strings

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs [hunk 5]
```diff
@@ -177,6 +178,110 @@ internal static unsafe HttpMethod GetKnownMethod(byte* data, int length, out int
...
 
+        /// <summary>
+        /// Parses string <paramref name="value"/> for a known HTTP method.
+        /// </summary>
+        /// <remarks>
+        /// A "known HTTP method" can be an HTTP method name defined in the HTTP/1.1 RFC.
+        /// The Known Methods (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE)
+        /// </remarks>
+        /// <returns><c>true</c> if the input matches a known string, <c>false</c> otherwise.</returns>
+        public static bool GetKnownMethod(string value, out HttpMethod method)
```
Comment (JamesNK) [inline src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Probably should be called `TryGetKnownMethod` since it is following the try pattern of returning a bool with an out param.

File: src/Kestrel.Core/Internal/Http/HttpProtocol.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.cs]: Had to change `None = byte.MaxValue` as so indexing and hashing uses the values

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs (hunk 5 above)
Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Changed

**Comment by benaadams** [comment]
K, done - I think

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs [hunk 6]
```diff
@@ -135,17 +136,17 @@ public static string GetAsciiStringEscaped(this Span<byte> span, int maxChars)
...
             {
-                method = GetKnownMethod(data, span.Length, out length);
+                method = TryGetKnownMethod(data, span.Length, out length);
                 return method != HttpMethod.Custom;
             }
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal static unsafe HttpMethod GetKnownMethod(byte* data, int length, out int methodLength)
+        internal static unsafe HttpMethod TryGetKnownMethod(byte* data, int length, out int methodLength)
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: This shouldn't be Try

Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Because it always returns a value in `out`?

Reverted

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs (hunk 5 above)
Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Reverted, always returns value in `out`

**Comment by benaadams** [comment]
K, now done... hopefully :)

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs (hunk 5 above)
Comment (davidfowl) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: What's the bool for then?

Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Ah good point can clean this up; return the string

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs [hunk 7]
```diff
@@ -177,6 +178,87 @@ internal static unsafe HttpMethod GetKnownMethod(byte* data, int length, out int
...
 
+        /// <summary>
+        /// Parses string <paramref name="value"/> for a known HTTP method.
+        /// </summary>
+        /// <remarks>
+        /// A "known HTTP method" can be an HTTP method name defined in the HTTP/1.1 RFC.
+        /// The Known Methods (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE)
+        /// </remarks>
+        /// <returns><c>true</c> if the input matches a known string, <c>false</c> otherwise.</returns>
+        public static string GetKnownMethod(string value, out HttpMethod method)
```
Comment (JamesNK) [inline src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: This method is taking string `value` and then returning it back again. Why not just return HttpMethod instead of an out param?

Also XML docs need to be updated.

Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Done

File: src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs [hunk 8]
```diff
@@ -90,8 +90,21 @@ string IHttpRequestFeature.Scheme
...
+                {
+                    return _methodText;
+                }
+
+                _methodText = HttpUtilities.MethodToString(Method) ?? string.Empty;
+                return _methodText;
+            }
+            set
+            {
+                Method = HttpUtilities.GetKnownMethod(value);
```
Comment (halter73) [inline src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: Out of curiosity, what would happen if we treated every request method set by the app via IHttpRequestFeature.Method as a custom method whether or not it known in order to elide this call to GetKnownMethod?

I don't see how this could be a problem, because none of Kestrel's logic should be predicated on the request method set by the application.

Comment (benaadams) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: So the app couldn't change request behavior based on method? If you changed method from `GET` -> `HEAD` the would mean if the app produced no body, but a content length it would throw with an 500 error.

However that may be better than responding with a `HEAD` response to a `GET` request and then hanging the browser while waiting for data?

Comment (Tratcher) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: While the app should be able to change all of the properties of HttpContext to lie to itself, the server must maintain a source of truth so it can properly interact with the client.

Comment (benaadams) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: Changed

**Comment by benaadams** [comment]
Changed and rebased

File: src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs (hunk 8 above)
Comment (benaadams) [reply src/Kestrel.Core/Internal/Http/HttpProtocol.FeatureCollection.cs]: This change is really messing up the tests :(

**Comment by benaadams** [comment]
Rebased

**Comment by halter73** [comment]
@pakrym you want to take a look?

**Comment by davidfowl** [comment]
@benaadams can you rebase?

File: src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs [hunk 9]
```diff
@@ -177,6 +178,87 @@ internal static unsafe HttpMethod GetKnownMethod(byte* data, int length, out int
...
+        /// Parses string <paramref name="value"/> for a known HTTP method.
+        /// </summary>
+        /// <remarks>
+        /// A "known HTTP method" can be an HTTP method name defined in the HTTP/1.1 RFC.
+        /// The Known Methods (CONNECT, DELETE, GET, HEAD, PATCH, POST, PUT, OPTIONS, TRACE)
+        /// </remarks>
+        /// <returns><see cref="HttpMethod"/></returns>
+        public static HttpMethod GetKnownMethod(string value)
+        {
+            // Called by http/2 and if user code overrides and sets the Method on IHttpRequestFeature
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: This is only used for http2 now, not setting the method directly right?

Comment (benaadams) [reply src/Kestrel.Core/Internal/Infrastructure/HttpUtilities.cs]: Yes, will change comment

**Comment by benaadams** [comment]
Rebased, changed comment

**Comment by halter73** [comment]
Thanks!

**Comment by sebastienros** [comment]
7% improvement on plaintext running on our physical.

image

**Comment by benaadams** [comment]
😄  What does the `CF` mean on those graphs?

**Comment by sebastienros** [comment]
Old benchmark to analyze the cost of connection adapters.
 
Used here: https://github.com/aspnet/benchmarks/blob/dev/src/Benchmarks/Program.cs#L234
With this filter: https://github.com/aspnet/benchmarks/blob/dev/src/Benchmarks/PassthroughConnectionFilter.cs#L12

**Comment by halter73** [comment]
This really shows what a difference one extra string comparison per request makes. I think `(object.ReferenceEquals(context.ServerSetMethod, HttpMethods.Head))` would have also worked, but moving to enum comparisons was much cleaner.

Thanks @benaadams for finding and fixing this.

**Comment by benaadams** [comment]
Don't think its just down to this one as there were 5 PRs merged at the same time; but glad they had an effect 😄 

Use enum for method rather than string compares (#2294)
Sanitize and centralize exception throws (#2293)
Flatten exception handling (#2313) 
Speed up TryGetAsciiString (#1973)
Faster IFeatureCollection.Get<TFeature> (#2290)

**Comment by benaadams** [comment]
Also none of them in the area of recent regression, so everything still to play for 😉
